### PR TITLE
Backlog count by status and issue type kpi's

### PR DESCRIPTION
**Pull request by aksshriv1**

### CODE REVIEW

File: common/src/main/java/com/publicissapient/kpidashboard/common/repository/jira/JiraIssueRepository.java [hunk 1]
```diff
@@ -157,4 +158,6 @@ JiraIssue findTopByProcessorIdAndBasicProjectConfigIdAndTypeNameAndChangeDateGre
 	 */
 	List<JiraIssue> findByNumberInAndBasicProjectConfigId(List<String> numberIds, String basicProjectConfigId);
 
+	List<JiraIssue> findByBasicProjectConfigIdIn(String basicProjectConfigId);
```
Comment (Chittauri) [inline common/src/main/java/com/publicissapient/kpidashboard/common/repository/jira/JiraIssueRepository.java:161]: add java doc

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/enums/KPIExcelColumn.java [hunk 2]
```diff
@@ -232,6 +232,10 @@ public enum KPIExcelColumn {
 
 	DEFECT_COUNT_BY_STATUS_RELEASE("kpi141", Arrays.asList("Issue ID", "Issue Description", "Sprint Name", "Issue Type",
 			"Issue Status", "Root Cause", "Priority", "Assignee")),
+	BACKLOG_COUNT_BY_STATUS("kpi151", Arrays.asList("Issue ID", "Issue Description", "Issue Type",
```
Comment (Chittauri) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/enums/KPIExcelColumn.java:235]: excel column script has to added

File: customapi/src/test/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
+package com.publicissapient.kpidashboard.apis.jira.scrum.service;
```
Comment (Chittauri) [inline customapi/src/test/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImplTest.java:1]: copyright

File: customapi/src/test/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
+package com.publicissapient.kpidashboard.apis.jira.scrum.service;
```
Comment (Chittauri) [inline customapi/src/test/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImplTest.java:1]: copyright

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
+package com.publicissapient.kpidashboard.apis.jira.scrum.service;
```
Comment (Chittauri) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java:1]: copyright

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+			Map<String, Integer> statusWiseCountMap) {
+		for (Map.Entry<String, List<JiraIssue>> statusEntry : statusData.entrySet()) {
+			statusWiseCountMap.put(statusEntry.getKey(), statusEntry.getValue().size());
+		}
+	}
+
+	private void projectWiseLeafNodeValue(List<Node> leafNodeList, KpiElement kpiElement, KpiRequest kpiRequest) {
+
+		String requestTrackerId = getRequestTrackerId();
+		leafNodeList.sort((node1, node2) -> node1.getSprintFilter().getStartDate()
```
Comment (Chittauri) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java:94]: why do we need to sort the sprint filter?

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	@Override
+	public Map<String, Object> fetchKPIDataFromDb(List<Node> leafNodeList, String startDate, String endDate,
+			KpiRequest kpiRequest) {
+
+		Map<String, Object> resultListMap = new HashMap<>();
+		Node leafNode = leafNodeList.stream().findFirst().orElse(null);
+
+		if (leafNode != null) {
+			log.info("BackLog Count By Status kpi -> Requested project : {}", leafNode.getProjectFilter().getName());
+			String basicProjectConfigId = leafNode.getProjectFilter().getBasicProjectConfigId().toString();
```
Comment (Chittauri) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java:57]: in backlog we need to get those statuses which are not tagged to sprint, check other kpis implementation, there are no checks present like that. please confirm.

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
+package com.publicissapient.kpidashboard.apis.jira.scrum.service;
```
Comment (Chittauri) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImpl.java:1]: check comments on BacklogCountByStatusType, implement same on this class

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+		if (CollectionUtils.isNotEmpty(jiraIssues) && leafNode != null) {
+
+			log.info("Backlog Count By Issue Type -> request id : {} total jira Issues : {}", requestTrackerId,
+					jiraIssues.size());
+
+			Map<String, List<JiraIssue>> typeWiseIssuesList = jiraIssues.stream()
+					.collect(Collectors.groupingBy(JiraIssue::getTypeName));
+
+			log.info("typeWiseIssuesList ->  : {}", typeWiseIssuesList);
```
Comment (hirbabar) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImpl.java:114]: typeWiseIssuesList -> remove unnecessary logs

**Review by hirbabar — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/publicissapient/kpidashboard/common/repository/jira/JiraIssueRepository.java (hunk 1 above)
Comment (aksshriv1) [reply common/src/main/java/com/publicissapient/kpidashboard/common/repository/jira/JiraIssueRepository.java:161]: added

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/enums/KPIExcelColumn.java (hunk 2 above)
Comment (aksshriv1) [reply customapi/src/main/java/com/publicissapient/kpidashboard/apis/enums/KPIExcelColumn.java:235]: Script is added

File: customapi/src/test/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImplTest.java (hunk 3 above)
Comment (aksshriv1) [reply customapi/src/test/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImplTest.java:1]: added

File: customapi/src/test/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImplTest.java (hunk 4 above)
Comment (aksshriv1) [reply customapi/src/test/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImplTest.java:1]: added

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java (hunk 5 above)
Comment (aksshriv1) [reply customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java:1]: added

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java (hunk 6 above)
Comment (aksshriv1) [reply customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java:94]: There is no need of this so removing this piece of code.

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImpl.java (hunk 9 above)
Comment (aksshriv1) [reply customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByIssueTypeServiceImpl.java:114]: removed

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java (hunk 7 above)
Comment (aksshriv1) [reply customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java:57]: I have verified with Sumit and harsh and as per them all jira issue are considered in the backlog

[the remark this answers is not in the document]
Comment (aksshriv1) [reply customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/BackLogCountByStatusServiceImpl.java:57]: Not possible as other KPI

File: database-scripts/scripts/kpi_master_insert.js [hunk 10]
```diff
@@ -2959,6 +2959,54 @@ db.getCollection('kpi_master').insert(
...
+  {
+      "kpiId": "kpi151",
+      "kpiName": "Backlog Count By Status",
+      "kpiUnit": "Count",
+      "isDeleted": "False",
+      "defaultOrder": 8,
+      "kpiCategory": "Backlog",
+      "kpiSource": "Jira",
+      "groupId": 10,
+      "thresholdValue": "",
```
[the remark this answers is not in the document]
Comment (aksshriv1) [reply database-scripts/scripts/kpi_master_insert.js:2972]: done

**Review by hirbabar — APPROVED** [review]

**Review by Chittauri — APPROVED** [review]

**Pull request merged**